frontend/src/formdescriptioneditor/FormDescriptionEditorWebSocketContainer.tsx

```diff
@@ -0,0 +1,271 @@
+    const id = event.dataTransfer.getData('text/plain');
+    setState((prevState) => {
+      const newWidget: Widget = {
+        id: uuid(),
+        label: id,
+        kind: isKind(id) ? id : 'Textfield',
+        children: [],
+      };
+      const widgets = [...prevState.widgets];
+      let index = widgets.indexOf(widget);
+      if (index === -1) {
+        widgets.push(newWidget);
+      } else {
+        widgets.splice(index, 0, newWidget);
+      }
+      const newState: FormDescriptionEditorWebSocketContainerState = { ...prevState, widgets };
+      return newState;
+    });
+  };
+  const handleDropInside = (event: React.DragEvent<HTMLDivElement>, widget: Widget) => {
+    const id = event.dataTransfer.getData('text/plain');
+    setState((prevState) => {
+      const newWidget: Widget = {
+        id: uuid(),
+        label: id,
+        kind: isKind(id) ? id : 'Textfield',
+        children: [],
+      };
+      const widgets = [...prevState.widgets];
+      widgets.find((parentWidget) => parentWidget.id === widget.id)?.children.push(newWidget);
+      const newState: FormDescriptionEditorWebSocketContainerState = { ...prevState, widgets };
+      return newState;
+    });
+  };
+
+  return (
+    <div className={classes.formDescriptionEditor}>
+      <div className={classes.header}>
+        <Typography>Form</Typography>
+      </div>
+      <div className={classes.main}>
+        <div className={classes.widgets}>
+          <Typography gutterBottom>Widgets</Typography>
+          <div id="Textfield" draggable="true" className={classes.widgetKind} onDragStart={handleDragStart}>
+            <TextFieldsIcon />
+            <Typography variant="caption" gutterBottom>
+              Textfield
+            </Typography>
+          </div>
+        </div>
+        <div className={classes.preview}>
+          <Typography>Preview</Typography>
+          <div className={classes.body}>
+            {widgets.map((widget) => (
+              <WidgetEntry
+                key={widget.id}
+                widget={widget}
+                selection={selection}
+                setSelection={setSelection}
+                onDropBefore={handleDropBefore}
+                onDropInside={handleDropInside}
+              />
+            ))}
+            <div
+              className={classes.bottomDropArea}
+              onDragEnter={handleDragEnter}
+              onDragOver={handleDragOver}
+              onDragLeave={handleDragLeave}
+              onDrop={handleDrop}
+            />
+          </div>
+        </div>
+      </div>
+    </div>
+  );
+};
+
+const useWidgetEntryStyles = makeStyles((theme) => ({
+  widget: {
+    display: 'flex',
+    flexDirection: 'column',
+    alignItems: 'stretch',
+  },
+  placeholder: {
+    height: '8px',
+  },
+  dragOver: {
+    border: 'dashed 1px red',
+  },
+}));
+
+const WidgetEntry = ({ widget, selection, setSelection, onDropBefore, onDropInside }: WidgetEntryProps) => {
```

Comment:
Contrary to my prototype, you do not need to store all the components in one file.

Reply:
Done